recommender/engine_test.go

```diff
@@ -43,35 +55,461 @@ func TestNewEngine(t *testing.T) {
 	}
 }
 
+// utility VmRegistry for mocking purposes
+type DummyVmRegistry struct {
+	// test case id to drive the behaviour
+	TcId int
+	VmId int
+}
+
+func (dvmr DummyVmRegistry) getAvailableAttributeValues(attr string) ([]float64, error) {
+	switch dvmr.TcId {
+	case 1:
+		// 3 values between 10 - 20
+		return []float64{12, 13, 14}, nil
+	case 2:
+		// 2 values between 10 - 20
+		return []float64{8, 13, 14, 6}, nil
+	case 3:
+		// no values between 10-20, return the closest value
+		return []float64{30, 40, 50, 60}, nil
+	case 4:
+		// no values between 10-20, return the closest value
+		return []float64{1, 2, 3, 5, 9}, nil
+	case 5:
+		// error, min > max
+		return []float64{1}, nil
+	case 6:
+		// error returned
+		return nil, errors.New("")
+
+	}
+
+	return nil, nil
+}
+
+func (dvmr DummyVmRegistry) findVmsWithAttrValues(region string, zones []string, attr string, values []float64) ([]VirtualMachine, error) {
+	if region == "errorRegion" {
```

Comment:
you can use the existing TcId here to drive the behaviour; please remove the extra field and use the switch statement only